featurebyte/service/column_attributes.py

```diff
@@ -0,0 +1,157 @@
+"""Module with extra services for data type info detection like embedding columns or flat (non-nested) dicts"""
+from __future__ import annotations
+
+from typing import Any
+
+import json
+from abc import ABC, abstractmethod
+
+import numpy as np
+import pandas as pd
+
+from featurebyte.api.source_table import SourceTable
+from featurebyte.common.utils import dataframe_from_json
+from featurebyte.enum import ColumnAttribute, DBVarType
+from featurebyte.logging import get_logger
+from featurebyte.models.feature_store import TableModel
+from featurebyte.query_graph.model.column_info import ColumnInfo
+from featurebyte.schema.feature_store import FeatureStoreSample
+from featurebyte.service.feature_store import FeatureStoreService
+from featurebyte.service.preview import PreviewService
+
+logger = get_logger(__name__)
+
+SAMPLE_SIZE = 1000
+RANDOM_SEED = 42
+
+
+class ColumnAttributesDetectionService:
+    """Service that detects attributes of the column with specific datatypes.
+    For example:
+        - Checking if ARRAY column has same number of dimensions in all observations (square),
+          and might represent embedding of the text.
+        - Checking if OBJECT is flat dictionary (no nested dictionaries).
+    """
+
+    def __init__(
+        self,
+        preview_service: PreviewService,
+        feature_store_service: FeatureStoreService,
+    ):
+        self.preview_service = preview_service
+        self.feature_store_service = feature_store_service
+
+        self.detectors: list[BaseColumnAttributesDetector] = [
+            ArrayEmbeddingColumnAttributesDetector(),
+            FlatDictColumnAttributesDetector(),
+        ]
+
+    async def add_columns_attributes(self, table: TableModel) -> None:
+        """Detect and adds columns attributes like embedding, flat dict, etc.
+
+        Parameters
+        ----------
+        table : TableModel
+            Table with columns to run detection for.
+        """
+        feature_store = await self.feature_store_service.get_document(
+            document_id=table.tabular_source.feature_store_id,
+        )
+        source_table = SourceTable(
+            feature_store=feature_store,
+            tabular_source=table.tabular_source,
+            columns_info=table.columns_info,
+        )
+        graph, node = source_table.frame.extract_pruned_graph_and_node()
```

Comment:
We could project those affected columns only (for example columns with `ARRAY` or `OBJECT` type).

Reply:
Done